feat: improved metric selector [DET-4122]

armandmcqueen opened the pull request:
## Description

Improve the metric selector. 

Allow users to select all metrics. Allow users to select all metrics that match the filter. Allow users to deselect all metrics. Improve placeholders to be more informative about how many metrics are selected and how many metric are available in total.


## Test Plan

<!---
Describe the situations in which you've tested your change, and/or a screenshot as appropriate.
Reviewers may ask questions of this test plan to ensure adequate manual coverage of changes.
-->


## Commentary (optional)

User facing:

- Added ALL option to dropdown
- While filtering, have ALL option select all options that meet the filter. 
- Adjust the appearance of the ALL option when filtering (when no filter = `All`, when filter has filtered down to N results = `All N results`, when filter has only a single result = do not display ALL option)
- Add a clear icon (antd out-of-the-box) that deselects all options. Upgrade antd to access the `onClear` callback to enable this behavior.
- Changed the text in the dropdown when it is not open. When all N options are selected, it is 'All N selected'. When X of the N options are selected, it is 'X of N selected'. When no options are selected, it is 'None selected'

Internal Notes:
- antd doesn't give us access to the string used to filter or how many results are shown after the filter, so we need to internally track the filter and the number of results that match so we can adjust the ALL option appearance.

---

@hkang1  @vishnu2kmohan 
- TODO: Do we want to be able to search for 'validation' and select all validation metrics? We don't currently include the metric type during the filtering, but we easily can.
- TODO: Should any of the Placeholder strings be improved?




## Checklist


- [x] Release notes should be added as a separate file under `docs/release-notes/`.

armandmcqueen commented on webui/react/src/components/MetricSelectFilter.tsx:
```diff
@@ -62,16 +87,91 @@ const MetricSelectFilter: React.FC<Props> = ({ metricNames, multiple, onChange,
     (onChange as MultipleHandler)(newMetric.sort(metricNameSorter));
   }, [ multiple, onChange, value ]);
 
+  const handleFiltering = useCallback((search: string, option) => {
```
This doesn't need useCallback, does it?

hkang1 replied:
This one does because the callback is being sent to the child component of `SelectFilter`. If we don't use it for `handleFiltering`, anytime `MetricSelectFilter` re-renders, `SelectFilter` will re-render as well because `handleFiltering` will be recreated without the `useCallback`.

armandmcqueen replied:
Ah, that makes sense, thanks.

armandmcqueen commented on webui/react/src/utils/trial.ts:
```diff
@@ -43,6 +43,26 @@ export const metricNameToValue = (metricName: MetricName): string => {
+      type: MetricType.Validation,
+    };
+  } else {
+    console.error("metricNameFromValue was called, but the metricName doesn't appear to " +
```
ESLint rule says we shouldn't use the console. This case could occur if we add new types of metrics in the future but forget to update this function. How do we make sure we notice if that happens?

hkang1 replied:
we have a general `handleError` function in `ErrorHandler.ts`. One common way we've been using this is calling `handleError` with the `silent` object property to be `true` if `process.env.IS_DEV` is `true`. Meaning, if this error occurs while in the dev environment it will present it to the user via notification. Meanwhile in the background it will send it to segments as part of error analytics if telemetry has been enabled.

hkang1 commented:
Love the changes, the `X of Total` label and the `All` option!

**suggestion:** I think the search input clear button (in the screenshot below) is a bit overloaded in terms of functionality. I noticed that it acts as the `None` option. I'm a little bit worried that users might expect that button to simply clear out the search input field instead of also deselecting all options. What if we keep the clear button just as an input clear and add the option `Reset` right below `All` and that will reset to selected the config search metric only (that way we don't run into the issue of having 0 metrics selected)?

![2020-10-09 at 5 33 PM](https://user-images.githubusercontent.com/220971/95639416-94360c00-0a55-11eb-9622-0ddb26eb58dc.png)

hkang1 commented on webui/react/src/components/MetricSelectFilter.tsx:
```diff
@@ -37,9 +39,31 @@ const MetricSelectFilter: React.FC<Props> = ({ metricNames, multiple, onChange,
+    return metricNames.length;
+  }, [ metricNames ]);
+
+  const filterFn = (search: string, metricName: string) => {
+    return metricName.toLocaleLowerCase().indexOf(search.toLocaleLowerCase()) !== -1;
+  };
```
**chore:** Since this function has no dependencies on anything inside the `MetricSelectFilter`, it can be moved just above the `const MetricSelectFilter: React...` line. This way it won't get recreated each time `MetricSelectFilter` rerenders.

hkang1 commented on webui/react/src/components/MetricSelectFilter.tsx:
```diff
@@ -37,9 +39,31 @@ const MetricSelectFilter: React.FC<Props> = ({ metricNames, multiple, onChange,
     return metricNames.filter(metric => metric.type === MetricType.Validation);
   }, [ metricNames ]);
 
+  const totalNumMetrics = useMemo(() => {
+    return metricNames.length;
+  }, [ metricNames ]);
```
**nit:** this can just one line:
```
const totalNumMetrics = useMemo(() => metricNames.length, [ metricNames ]);
```

hkang1 commented:
**suggestion:** just noting here on adding a default metric prop for the `Reset` behavior.

armandmcqueen commented on webui/react/src/utils/trial.ts:
```diff
@@ -43,6 +45,45 @@ export const metricNameToValue = (metricName: MetricName): string => {
+    const errName = 'metricNameFromValueUnrecognizedMetricType';
+    const errSlug = 'metricnamefromvalue-unrecognized-metric-type';
+    const errMessage = `metricNameFromValue was called, but the metricName doesn't appear to " +
+        'be a training metric or a validation metric (${metricValue})`;
```
@hkang1  Is including the `metricValue` as part of the error message too much detail?

hkang1 replied:
I think `metricValue` would be great to have since this is more for us to use when figuring out how we got something that is neither training nor validation metric.

hkang1 replied:
**suggestion:** I think in template literals, everything will be preserved in the string, including the double quote, plus character and the single quote. You can do something like this and the newlines are typically not a problem unless we are presenting them in an element with a CSS definition of `white-space: pre;`, which is really rare and something we avoid in general.
```
const errMessage = `
  metricNameFromValue was called, but the metricName doesn't appear to
  be a training metric or a validation metric (${metricValue})
`;
```

armandmcqueen replied:
Good catch! I changed it from a normal string to a template string and didn't notice that the quotes/plus sign became part of the string. Fixed

armandmcqueen commented:
I added a Reset option and changed how we log the error. 

I also removed the clear button because it is ambiguous. When the dropdown is open, it can look like it should clears the input. When the dropdown is closed, it looks like it should clear the selections.

hkang1 commented on webui/react/src/utils/trial.ts:
```diff
@@ -43,6 +45,45 @@ export const metricNameToValue = (metricName: MetricName): string => {
+      isUserTriggered: false,
+      level: ErrorLevel.Error,
+      message: errMessage,
+      silent: silent,
```
**nit:** you can shorten the silent definition by doing something like this:
```
  silent: !process.env.IS_DEV,
```

armandmcqueen replied:
fixed

hkang1 commented:
The labels and the options all look and work fantastic!

Some additional suggestions after playing around with this:
- Change `Reset To Default` to `Reset to Default`
- When the user clicks `All` or `Reset to Default`, perform the selection and then dismiss the dropdown instead of leaving it open.

hkang1 reviewed (commented):
The metric select filter is looking really sharp and a lot of powerful features built into this one area!

I have some minor comments and two suggestions. Hoping they are easy to do...

armandmcqueen commented:
- Fixed the option to be `Reset to Default`. 
- Dropdown now closes after selecting `All` or `Reset` option. 
- Dropdown does not close and clear the filter when selecting any other option.

hkang1 commented on webui/react/src/components/MetricSelectFilter.tsx:
```diff
@@ -37,9 +52,30 @@ const MetricSelectFilter: React.FC<Props> = ({ metricNames, multiple, onChange,
   const handleMetricSelect = useCallback((newValue: SelectValue) => {
     if (!onChange) return;
 
+    if ((newValue as string) === allOptionId) {
+      (onChange as MultipleHandler)(visibleMetrics.sort(metricNameSorter));
+      selectRef.current?.blur();
+      return;
+    }
+    if ((newValue as string) === resetOptionId) {
+      (onChange as MultipleHandler)(defaultMetricNames.sort(metricNameSorter));
+      selectRef.current?.blur();
+      return;
+    }
+
```
👍 worked like a charm!

hkang1 commented on webui/react/src/components/MetricSelectFilter.tsx:
```diff
@@ -62,16 +98,89 @@ const MetricSelectFilter: React.FC<Props> = ({ metricNames, multiple, onChange,
+      return true;
+    }
+    if (!option.value) {
+      // Handle optionGroups that don't have a value to make TS happy. They aren't
+      // impacted by filtering anyway
```
**nit:** For multi-line comments we use the following format to make them more prominent.
```
/*
 * line 1
 * line 2
 */
```
Used to have a linter do this but it made it difficult to comment and uncomment code.

hkang1 commented on webui/react/src/components/MetricSelectFilter.tsx:
```diff
@@ -62,16 +98,89 @@ const MetricSelectFilter: React.FC<Props> = ({ metricNames, multiple, onChange,
+    onSearch={handleSearchInputChange}
     onSelect={handleMetricSelect}>
+
+    { multiple && visibleMetrics.length > 0 &&
+    <Option key={resetOptionId} value={resetOptionId}>
+      <BadgeTag label='Reset to Default' />
+    </Option>}
+
+    { multiple && visibleMetrics.length > 1 && allOption}
```
**question:** Should we make these persist (without a dependency on `visibleMetrics.length`)? Where I found this to be a little confusing is when I was changing the filtering text, it dynamically added and removed the `Rest to Default` and `All` options.

hkang1 reviewed (approved):
LGTM, Nice work on this PR! One small non-blocking nit around comment format that the linter doesn't catch.
